### PR TITLE
[TRTLLM-6825][fix] Update lora for phi4-mm

### PR DESCRIPTION
**Pull request by Wanli-Jiang**
## Summary by CodeRabbit

* **Refactor**
  * LoRA target renamed: mlp_h_to_4h → mlp_gate_up and mapping aligned to gate_up_proj.
  * Removed deprecated lora_dir from LoRA setup, simplifying configuration.
* **New Features**
  * Added a configurable swap_gate_up_proj flag (defaults to preserve prior behavior) to control gate_up_proj weight swapping.
* **Tests**
  * Updated integration test expectations and configs to match the mlp_gate_up change and new swap flag.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Review by yechank-nvidia — APPROVED** [review]
Approving as modeling reviewer. Seems needing approval from peft devs.

File: tests/integration/defs/test_e2e.py [hunk 1]
```diff
@@ -2395,15 +2395,15 @@ def test_ptp_quickstart_multimodal_phi4mm(llm_root, llm_venv, modality):
     }
     expected_keywords = {
         "image": [
-            ["image", "depicts", "mountain", "half", "rock"],
-            ["road", "car", "lane", "traffic", "bus"],
+            ["object", "mountain", "weather", "clear", "clouds"],
```
[the remark this answers is not in the document]
Comment (Wanli-Jiang) [reply tests/integration/defs/test_e2e.py:2398]: Before PR, the output is also meaningful, since LoRA just tuned weights a bit.

We cannot find issue from the generated response. We found issue from the image-input accuracy bench.

File: tensorrt_llm/lora_manager.py [hunk 2]
```diff
@@ -243,6 +243,7 @@ class LoraModelConfig:
     trtllm_modules_to_hf_modules: dict[str, str]
     hidden_size: int
     dtype: str
+    swap_gate_up_proj: bool = True
```
[the remark this answers is not in the document]
Comment (Wanli-Jiang) [reply tensorrt_llm/lora_manager.py]: Left comments

File: tensorrt_llm/lora_helper.py [hunk 3]
```diff
@@ -88,6 +88,10 @@ class LoraConfig(DictConversion):
     trtllm_modules_to_hf_modules: Dict[str, str] = field(default_factory=dict)
     max_loras: Optional[int] = None
     max_cpu_loras: Optional[int] = None
+    # swap gate_up_proj.lora_B.weight or not during loading LoRA checkpoint.
+    # Default is True to keep code consistency.
+    # Phi4-mm model don't need to swap gate_up_proj.lora_B.weight.
+    swap_gate_up_proj_lora_b_weight: bool = True
```
Comment (amitz-nv) [inline tensorrt_llm/lora_helper.py:94]: I'm not familiar with phi4mm, can you explain how come it doesn't need this swap? or how did you find out this is an issue with this model?
I want to be sure I understand the necessity of this new field

File: tensorrt_llm/lora_helper.py [hunk 4]
```diff
@@ -88,6 +88,10 @@ class LoraConfig(DictConversion):
     trtllm_modules_to_hf_modules: Dict[str, str] = field(default_factory=dict)
     max_loras: Optional[int] = None
     max_cpu_loras: Optional[int] = None
+    # swap gate_up_proj.lora_B.weight or not during loading LoRA checkpoint.
+    # Default is True to keep code consistency.
+    # Phi4-mm model don't need to swap gate_up_proj.lora_B.weight.
```
Comment (amitz-nv) [inline tensorrt_llm/lora_helper.py:93]: IMO these comments may be removed

File: tensorrt_llm/lora_manager.py [hunk 5]
```diff
@@ -989,7 +990,10 @@ def load_from_model_dir(uid, model_dir, hf_config):
             lora_model = load_state_dict(get_model_path(model_dir, "adapter_model"))
             if lora_model is None:
                 raise ValueError(f"Failed to load adapter_model from {model_dir}")
-            lora_model = preprocess_lora_weights(lora_model)
+            swap_gate_up_proj_lora_b_weight = getattr(
+                model_config, "swap_gate_up_proj_lora_b_weight", True
+            )
+            lora_model = preprocess_lora_weights(lora_model, swap_gate_up_proj_lora_b_weight)
```
[the remark this answers is not in the document]
Comment (Wanli-Jiang) [reply tensorrt_llm/lora_manager.py]: done

File: tensorrt_llm/lora_helper.py (hunk 4 above)
Comment (Wanli-Jiang) [reply tensorrt_llm/lora_helper.py]: done

**Review by amitz-nv — APPROVED** [review]
LGTM, please remove the comment lines in `LoraConfig` before merge

**Review by amukkara — APPROVED** [review]

**Review by venkywonka — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 6]
```diff
@@ -459,13 +459,16 @@ def __init__(
     def runtime_draft_len(self):
         return self.max_draft_len if self.enable_spec_decode else 0
 
-    def set_lora_model_config(self, lora_target_modules: list[str],
-                              trtllm_modules_to_hf_modules: dict[str, str]):
+    def set_lora_model_config(self,
+                              lora_target_modules: list[str],
+                              trtllm_modules_to_hf_modules: dict[str, str],
+                              swap_gate_up_proj: bool = True):
```
[the remark this answers is not in the document]
Comment (Wanli-Jiang) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: I set it True to keep code consistency, otherwise other models with Lora will change their behaviours.

**Review by yuxianq — APPROVED** [review]

**Review by shaharmor98 — COMMENTED** [review]
Overall looks good.  
I have one major comment though.

**Review by shaharmor98 — APPROVED** [review]
LGTM.
As agreed with Wanli, I'm approving this PR to unblock the current release branch. A subsequent PR with some minor design changes is expected to be opened.
Thanks for the cooperation and good work.

File: tensorrt_llm/lora_helper.py (hunk 4 above)
Comment (shaharmor98) [reply tensorrt_llm/lora_helper.py]: @amitz-nv this will be changed in a subsequent PR
